res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -0,0 +1,890 @@
+        //       of each physical deck component to determine which deck it's manipulating in
+        //       Mixxx. The deck load buttons, however, are both on MIDI channel 16, so I need
+        //       a better way to determine this.
+        off: Prime4.rgbCode.whiteDark,
+        on: Prime4.rgbCode.white,
+    });
+
+    this.reconnectComponents(function(c) {
+        if (c.group === undefined) {
+            c.group = this.currentDeck;
+        }
+    });
+};
+
+Prime4.Deck.prototype = new components.Deck();
+
+Prime4.shift = false;
+Prime4.shiftState = function(channel, control, value) {
+    Prime4.shift = value === 0x7F;
+    if (Prime4.shift) {
+        midi.sendShortMsg(0x90 + channel, control, 0x02);
+        Prime4.leftDeck.shift();
+        Prime4.rightDeck.shift();
+        Prime4.effectBank[0].shift();
+        Prime4.effectBank[1].shift();
+    } else {
+        midi.sendShortMsg(0x90 + channel, control, 0x01);
+        Prime4.leftDeck.unshift();
+        Prime4.rightDeck.unshift();
+        Prime4.effectBank[0].unshift();
+        Prime4.effectBank[1].unshift();
+    }
+};
+
+//========== PERFORMANCE PADS ==========//
+
+// Access the appropriate mode-select pad without remembering MIDI values
+Prime4.padMode = {
+    HOTCUE: 0x0B,
+    LOOP: 0x0C,
+    ROLL: 0x0D,
+    SLICER: 0x0E,
+};
+
+Prime4.PadSection = function(deck, offset) {
+    components.ComponentContainer.call(this);
+    const theContainer = this;
+
+    // Create component containers for each pad mode
+    const modes = new components.ComponentContainer({
+        "hotcue": new Prime4.hotcueMode(deck, offset),
+        "loop": new Prime4.loopMode(deck, offset),
+        //TODO: "autoloop": new Prime4.autoloopMode(deck, offset),
+        "roll": new Prime4.rollMode(deck, offset),
+        "sampler": new Prime4.samplerMode(deck, offset),
+    });
+
+    modes.forEachComponent(c => c.disconnect());
+
+    const controlToPadMode = control => {
+        let mode;
+        switch (control) {
+        case Prime4.padMode.HOTCUE:
+            mode = modes.hotcue;
+            break;
+        case Prime4.padMode.LOOP:
+            mode = modes.loop;
+            /*
+            // TODO: Implement autoloop mode
+            if (Prime4.PadSection.currentMode === modes.loop) {
+                mode = modes.autoloop;
+            } else {
+                mode = modes.loop;
+            }
+            */
+            break;
+        case Prime4.padMode.ROLL:
+            mode = modes.roll;
+            break;
+        case Prime4.padMode.SLICER:
+            mode = modes.sampler;
+            break;
+        }
+
+        return mode;
+    };
+
+    this.offset = offset;
+
+    this.padModeSelectLeds = new components.Component({
+        trigger: function() {
+            modes.forEach(mode => {
+                midi.sendShortMsg(0x94 + offset, mode.ledControl, theContainer.currentMode === mode ? mode.colourOn : mode.colourOff);
+            });
```

Comment:
```suggestion
            modes.forEachComponent(mode => {
                midi.sendShortMsg(0x94 + offset, mode.ledControl, theContainer.currentMode === mode ? mode.colourOn : mode.colourOff);
            }, false);
```

Reply:
Fixed the error, but the `mode.ledControl` and `mode.colourOn`/`mode.colourOff` parameters aren't working. Latest code is pushed, investigating the issue now. Pretty sure it's because `mode` isn't in this function's scope.

Reply:
mhmm. that should work as long as each padmode defines them. `mode` is in the function scope because its defined as the parameter of the arrow function. I'm not sure why its undefined. Can you verify that each padmode is a actually a componentcontainer?

```javascript
            modes.forEachComponent(mode => {
                if (!(mode instanceof components.ComponentContainer)) {
                    throw "padmode is not a container as expected";
                }
                midi.sendShortMsg(0x94 + offset, mode.ledControl, theContainer.currentMode === mode ? mode.colourOn : mode.colourOff);
            }, false);
```

Reply:
I can't see the `throw` message in the console, but every single LED on the controller is off except for the performance pads on the left side, so clearly something got thrown

Reply:
If I use `console.log` instead of `throw`, I can see the padmode-not-container message every time the function gets triggered

Reply:
When I start Mixxx, the message seems to get printed hundreds of times, I can't count them all.

Each time it gets printed, it comes with this error log in the console:

```
 js:debug [Controller] padmode is not a container as expected
warning [Controller] "Could not convert argument 1 at"
warning [Controller] 	 "@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:736"
warning [Controller] 	 "applyOperationTo@file:///home/whanake/Git/mixxx/res/controllers/midi-components-0.0.js:532"
warning [Controller] 	 "forEachComponent@file:///home/whanake/Git/mixxx/res/controllers/midi-components-0.0.js:544"
warning [Controller] 	 "applyOperationTo@file:///home/whanake/Git/mixxx/res/controllers/midi-components-0.0.js:534"
warning [Controller] 	 "forEachComponent@file:///home/whanake/Git/mixxx/res/controllers/midi-components-0.0.js:544"
warning [Controller] 	 "applyOperationTo@file:///home/whanake/Git/mixxx/res/controllers/midi-components-0.0.js:534"
warning [Controller] 	 "forEachComponent@file:///home/whanake/Git/mixxx/res/controllers/midi-components-0.0.js:544"
warning [Controller] 	 "trigger@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:731"
warning [Controller] 	 "@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:768"
warning [Controller] 	 "@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:772"
warning [Controller] 	 "@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:512"
warning [Controller] 	 "@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:267"
warning [Controller] "Passing incompatible arguments to C++ functions from JavaScript is dangerous and deprecated."
warning [Controller] "This will throw a JavaScript TypeError in future releases of Qt!"
warning [Controller] "Could not convert argument 2 at"
warning [Controller] 	 "@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:736"
warning [Controller] 	 "applyOperationTo@file:///home/whanake/Git/mixxx/res/controllers/midi-components-0.0.js:532"
warning [Controller] 	 "forEachComponent@file:///home/whanake/Git/mixxx/res/controllers/midi-components-0.0.js:544"
warning [Controller] 	 "applyOperationTo@file:///home/whanake/Git/mixxx/res/controllers/midi-components-0.0.js:534"
warning [Controller] 	 "forEachComponent@file:///home/whanake/Git/mixxx/res/controllers/midi-components-0.0.js:544"
warning [Controller] 	 "applyOperationTo@file:///home/whanake/Git/mixxx/res/controllers/midi-components-0.0.js:534"
warning [Controller] 	 "forEachComponent@file:///home/whanake/Git/mixxx/res/controllers/midi-components-0.0.js:544"
warning [Controller] 	 "trigger@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:731"
warning [Controller] 	 "@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:768"
warning [Controller] 	 "@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:772"
warning [Controller] 	 "@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:512"
warning [Controller] 	 "@file:///home/whanake/.mixxx/controllers/Denon-Prime-4-scripts.js:267"
warning [Controller] "Passing incompatible arguments to C++ functions from JavaScript is dangerous and deprecated."
warning [Controller] "This will throw a JavaScript TypeError in future releases of Qt!"
```

Reply:
When I press one of the pad-mode-switching buttons, the `console.log` and error message prints exactly 32 times.